src/core/math_func.hpp

```diff
@@ -124,6 +124,57 @@ static inline uint ClampU(const uint a, const uint min, const uint max)
 	return Clamp<uint>(a, min, max);
 }
 
+/**
+ * Clamp the given value down to lie within the requested type.
+ *
+ * For example ClampTo<uint8_t> will return a value clamped to the range of 0
+ * to 255. Anything smaller will become 0, anything larger will become 255.
+ *
+ * @param a The 64-bit value to clamp.
+ * @return The 64-bit value reduced to a value within the given allowed range
+ * for the return type.
+ * @see Clamp(int, int, int)
+ */
+template <typename To, typename From>
+constexpr To ClampTo(From value)
+{
+	if constexpr (!std::numeric_limits<To>::is_integer || !std::numeric_limits<From>::is_integer) {
+		return "Both From and To type must be integrals";
+	}
+
+	if constexpr (std::is_same<To, From>::value) {
+		/* If we clamp to the same type, just do nothing. */
+		return value;
+	}
+
+	if constexpr (std::numeric_limits<To>::is_signed) {
+		/* The output is a signed number. */
+		if constexpr (std::numeric_limits<From>::is_signed) {
+			/* Input is signed, so use int64_t as intermediate type. */
+			return static_cast<To>(std::clamp<int64_t>(value,
```

Comment:
Is std::clamp<From>() enough, instead of promoting to 64 bits?

Reply:
I think I found a solution, although it does make it a more difficult to understand.